packages/data-broker/src/services/dhis/DhisInputSchemeResolvingApiProxy.js

```diff
@@ -192,6 +185,63 @@ export class DhisInputSchemeResolvingApiProxy {
     return modifiedQuery;
   };
 
+  /**
+   * @param response {*}
+   * @param dataElementCodes {*}
+   * @returns {*}
+   * @private
+   */
+  translateDataElementIdsToCodesInResponse = async (response, dataElementCodes) => {
+    const dataElementIdToCode = {};
+
+    const dataElements = await this.models.dataSource.find({
+      code: dataElementCodes,
+      type: 'dataElement',
+    });
+
+    for (const dataElement of dataElements) {
+      if (dataElement.config.dhisId) {
+        dataElementIdToCode[dataElement.config.dhisId] = dataElement.code;
+      }
+    }
+
+    return translateElementKeysInEventAnalytics(response, dataElementIdToCode);
+  };
+
+  /**
+   * @param response {*}
+   * @returns {*}
+   * @private
+   */
+  translateOrgUnitIdsToCodesInResponse = async response => {
+    if (!response.rows.length) return response;
+    const newRows = [];
+
+    const orgUnitIdIndex = response.headers.findIndex(({ name }) => name === 'ou');
+    const orgUnitCodeIndex = response.headers.findIndex(({ name }) => name === 'oucode');
+    if (!orgUnitIdIndex || !orgUnitCodeIndex)
+      throw new Error("Can't read org unit id/code from dhis");
+
+    const dhisIds = response.rows.map(row => row[orgUnitIdIndex]);
+
+    const mappings = await this.models.dataServiceEntity.find({
+      [QUERY_CONJUNCTIONS.RAW]: {
+        sql: `config->>'dhis_id' in (${dhisIds.map(() => '?')})`,
+        parameters: dhisIds,
+      },
+    });
+
+    for (const row of response.rows) {
+      const newRow = [...row];
+      const dhisId = row[orgUnitIdIndex];
+      const mapping = mappings.find(m => m.config.dhis_id === dhisId);
```

Comment:
Probably more efficient to do a `reduceToDictionary` in advance, rather than having this inner loop here (likely to have many rows for each org unit)

Reply:
Great point!